splice_machine/src/main/java/com/splicemachine/derby/impl/sql/compile/MergeJoinStrategy.java

```diff
@@ -277,29 +283,37 @@ private boolean mergeable(RowOrdering outerRowOrdering,
                         "Programmer error: RelationalOperator of type "+ relop.getClass()+" detected";
                 BinaryRelationalOperatorNode bron = (BinaryRelationalOperatorNode)relop;
                 if (bron.getOperator() == RelationalOperator.EQUALS_RELOP) {
-                    ColumnReference innerColumn = relop.getColumnOperand(innerTable);
-                    ColumnReference outerColumn = getOuterColumn(bron, innerColumn);
-                    if (innerColumn == null || outerColumn == null) continue;
-                    int innerColumnNumber = innerColumn.getColumnNumber();
+                    ColumnReference innerColumn = null;
+                    int innerColumnNumber;
+                    int outerTableNum;
+                    int outerColNum;
+
+                    if (isIndexOnExpr) {
+                        innerColumnNumber = pred.hasEqualOnIndexExpression(innerTable);
+                    } else {
+                        innerColumn = relop.getColumnOperand(innerTable);
+                        if (innerColumn == null) {
+                            continue;
+                        }
+                        innerColumnNumber = innerColumn.getColumnNumber();
+                    }
+
+                    Pair<Integer, Integer> outerColumnInfo = getOuterColumnInfo(bron, innerTable, innerColumn);
+                    if (outerColumnInfo != null) {
+                        outerTableNum = outerColumnInfo.getFirst();
+                        outerColNum = outerColumnInfo.getSecond();
+                    } else {
+                        continue;
+                    }
```

Comment:
i would do a slightly reordering here:
```
if (outerColumnInfo == null) continuel

int outerTableNum = outerColumnInfo.getFirst();
int outerColNum = outerColumnInfo.getSecond();
```

that way, there's no doubt outerTableNum and outerColNum are initialized.

Reply:
Done.